### PR TITLE
[PIP-60] [Proxy-Client] Support SNI routing for Pulsar CPP client

### PR DESCRIPTION
**Pull request by rdhabalia**
### Motivation

#8957 was added incorrectly and it was not supporting SNI routing for CPP client library.
This PR adds support of SNI roting to Pulsar CPP client library and allows users to connect with Pulsar broker using SNI Proxy.
User can configure Pulsar client lib to use SNI routing and proxy by adding below configuration
```
const std::string lookupUrl = "pulsar+ssl://broker-url:6651";
const std::string proxyUrl = "pulsar+ssl://sni-proxy-url:6651";
const std::string topicName = "persistent://pulsar/stg-use1/ns1/t1";
const std::string certFile = "/client-cert.pem";
const std::string keyFile = "/client-key.pem";
const std::string trustFile = "/trust-store.pem";

ClientConfiguration conf;
conf.setUseTls(true);
conf.setTlsAllowInsecureConnection(false);
conf.setTlsTrustCertsFilePath(trustFile);
conf.setAuth(AuthTls::create(std::move(certFile), std::move(keyFile)));
conf.setProxyServiceUrl(proxyUrl);
conf.setProxyProtocol(ClientConfiguration::SNI);

Client client(lookupUrl, conf);
```

### Modifications

Added a support of SNI routing by adding new APIs to configure proxy URL and protocol.
```
ClientConfiguration::setProxyServiceUrl(..)
ClientConfiguration::setProxyProtocol(..)
```

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Documentation

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)

- [ ] `doc-not-needed` 
(Please explain why)

- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### CODE REVIEW

File: include/pulsar/ClientConfiguration.h [hunk 1]
```diff
@@ -320,6 +324,33 @@ class PULSAR_PUBLIC ClientConfiguration {
...
+     *
+     * SNI routing:
+     * https://docs.trafficserver.apache.org/en/latest/admin-guide/layer-4-routing.en.html#sni-routing.
+     *
+     * @param proxyProtocol possible options (SNI)
+     * @return
+     */
+    ClientConfiguration& setProxyProtocol(const ProxyProtocol proxyProtocol);
+
+    const ProxyProtocol getProxyProtocol() const;
```
Comment (BewareMyPower) [inline include/pulsar/ClientConfiguration.h:352]: It's better to return a non-const value type, see https://stackoverflow.com/questions/8716330/purpose-of-returning-by-const-value

File: include/pulsar/ClientConfiguration.h [hunk 2]
```diff
@@ -320,6 +324,33 @@ class PULSAR_PUBLIC ClientConfiguration {
...
+     * Set appropriate proxy-protocol along with proxy-service url. Currently Pulsar supports SNI proxy
+     * routing.
+     *
+     * SNI routing:
+     * https://docs.trafficserver.apache.org/en/latest/admin-guide/layer-4-routing.en.html#sni-routing.
+     *
+     * @param proxyProtocol possible options (SNI)
+     * @return
+     */
+    ClientConfiguration& setProxyProtocol(const ProxyProtocol proxyProtocol);
```
Comment (BewareMyPower) [inline include/pulsar/ClientConfiguration.h:350]:
```suggestion
    ClientConfiguration& setProxyProtocol(ProxyProtocol proxyProtocol);
```

`const` for the parameter only works in the method implementation. When providing an API for users, it's better only to use `const` for references and pointers.

**Review by BewareMyPower — COMMENTED** [review]
Overall LGTM except the `const` use in the method signatures.

File: include/pulsar/ClientConfiguration.h (hunk 2 above)
Comment (rdhabalia) [reply include/pulsar/ClientConfiguration.h]: addressed it.

File: lib/ClientConfiguration.cc [hunk 3]
```diff
@@ -134,6 +134,23 @@ ClientConfiguration& ClientConfiguration::setConcurrentLookupRequest(int concurr
...
 
+ClientConfiguration& ClientConfiguration::setProxyServiceUrl(const std::string& proxyServiceUrl) {
+    impl_->proxyServiceUrl = proxyServiceUrl;
+    return *this;
+}
+
+const std::string& ClientConfiguration::getProxyServiceUrl() const { return impl_->proxyServiceUrl; }
+
+ClientConfiguration& ClientConfiguration::setProxyProtocol(
+    ClientConfiguration::ProxyProtocol proxyProtocol) {
```
Comment (BewareMyPower) [inline lib/ClientConfiguration.cc:145]:
```suggestion
ClientConfiguration& ClientConfiguration::setProxyProtocol(ClientConfiguration::ProxyProtocol proxyProtocol) {
```

Fix format check

**Review by BewareMyPower — APPROVED** [review]

**Pull request merged**